### PR TITLE
added the type 'Optional' to all optional parameters for async functions

### LINKED ISSUE #854: Typing for optional function parameters are incorrect
#### Environment details

  - OS type and version: MacOS
  - Python version: 3.11
  - pip version: 23.3.1
  - `google-cloud-firestore` version: 2.14.0

#### Steps to reproduce

  1. Use a function that has an optional parameter
  2. Observe type error

#### Code example

```python
async_client = AsyncClient()
async_col_ref = async_client.collection("users")
all_users = async_col_ref.stream()
```

#### Stack trace
```
Argument of type "None" cannot be assigned to parameter "transaction" of type "Transaction" in function "stream"
  "None" is incompatible with "Transaction"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
```

### PR DESCRIPTION
**Pull request by dcgudeman**
This should fix the typing errors for async functions

Fixes #854

@daniel-sanche

### CODE REVIEW

**Comment by daniel-sanche** [comment]
Hey David, thanks for putting this together. 

At the root of the typing issues in this library is the fact that the mypy type checker is currently mis-configured, so none of the type hints are verified, and a lot of them are wrong. I don't want to make changes to the type hints without getting the verification check working, to make sure we're making the right changes. But turning it all on at once would force a giant change across the entire codebase. So I'm thinking we'll have to explicitly opt-in individual files as we touch different parts of the code, and get the types fixed over time.

TL;DR: Do you think you can change the [mypy session in the noxfile](https://github.com/googleapis/python-firestore/blob/3a572b5d92b7bb20232b0b271a457fc4b6572f0e/noxfile.py#L146) to validate the types of the files you changed?:
```
@nox.session(python=DEFAULT_PYTHON_VERSION)
def mypy(session):
    """Verify type hints are mypy compatible."""
    session.install("-e", ".")
    session.install("mypy", "types-setuptools", "types-protobuf")
    session.run("mypy", "--follow-imports=skip", "--ignore-missing-imports",
        "google/cloud/firestore_v1/async_aggregation.py",
        "google/cloud/firestore_v1/async_document.py",
        "google/cloud/firestore_v1/async_batch.py",
        "google/cloud/firestore_v1/async_client.py",
        "google/cloud/firestore_v1/async_collection.py",
        "google/cloud/firestore_v1/async_query.py",
    )u
```

It looks like with this mypy check running, there are 9 errors that still need to be addressed. If you're willing to take those on and get these files passing, that would be great! But if it looks like too much work, let me know and I can take a look

Thanks

**Comment by daniel-sanche** [comment]
closing this in favor of https://github.com/googleapis/python-firestore/pull/984
